dvc/remote/base.py

```diff
@@ -680,20 +680,49 @@ def safe_remove(self, path_info, force=False):
         self.remove(path_info)
 
     def _checkout_file(
-        self, path_info, checksum, force, progress_callback=None
+        self,
+        path_info,
+        checksum,
+        force,
+        progress_callback=None,
+        save_link=True,
     ):
-        cache_info = self.checksum_to_path_info(checksum)
-        if self.exists(path_info):
-            msg = "data '{}' exists. Removing before checkout."
-            logger.warning(msg.format(str(path_info)))
-            self.safe_remove(path_info, force=force)
+        # NOTE: In case if path_info is already cached and path_info's
+        # link type matches cache link type, we would like to avoid
+        # relinking.
+        if self.changed(
+            path_info, {self.PARAM_CHECKSUM: checksum}
+        ) or not self._is_same_link_as_cache(path_info):
+            self._remove_and_relink(path_info, checksum, force, save_link)
```

Comment:
It feels like these two functions are excessive. For example `remove_and_relink` is a strange name, as `relink` already means that you are removing something if needed and creating a new link. Maybe let's not add these wrapper functions and leave the meat in here?

Reply:
done, removed wrappers